File: serverScripts/coordinate_calculations.py
def get_build_coordinates(player_pos, player_angle, structure_dimensions):
	start_x = player_pos.x
	start_y = player_pos.y
	start_z = player_pos.z

	direction = get_general_direction(player_angle)

	if direction == "north":
		end_x, end_y, end_z = get_facing_north_coordinates(start_x, start_y, start_z, structure_dimensions)
		direction = direction + "north"
	elif direction == "east":
		end_x, end_y, end_z = get_facing_east_coordinates(start_x, start_y, start_z, structure_dimensions)
		direction = direction + "east"
	elif direction == "south":
		end_x, end_y, end_z = get_facing_south_coordinates(start_x, start_y, start_z, structure_dimensions)
		direction = direction + "south"
	else:
		end_x, end_y, end_z = get_facing_west_coordinates(start_x, start_y, start_z, structure_dimensions)
		direction = direction + "west"

	return start_x, start_y, start_z, end_x, end_y, end_z, direction


def get_hollow_dimensions(start_x, start_y, start_z, end_x, end_y, end_z, player_angle, structure_dimensions):
	direction = get_general_direction(player_angle)

	if structure_dimensions[1] >= 3:
		start_y += 1
		end_y -= 1

	# fix x and z coordinates
	if direction == "west":  # ok
		start_x -= 1
		end_x += 1
		start_z += 1
		end_z -= 1
	elif direction == "north":
		start_x -= 1
		end_x += 1
		start_z -= 1
		end_z += 1
	elif direction == "east":  # ok
		start_x += 1
		end_x -= 1
		start_z -= 1
		end_z += 1
	else:  # ok
		start_x += 1
		end_x -= 1
		start_z += 1
		end_z -= 1

	return start_x, start_y, start_z, end_x, end_y, end_z


def get_facing_north_coordinates(start_x, start_y, start_z, structure_dimensions):
	end_z = start_z - structure_dimensions[0] + 1
	end_y = start_y + structure_dimensions[1] - 1
	end_x = start_x - structure_dimensions[2] + 1
	return end_x, end_y, end_z


def get_facing_east_coordinates(start_x, start_y, start_z, structure_dimensions):  # fixed
	end_z = start_z - structure_dimensions[2] + 1
	end_y = start_y + structure_dimensions[1] - 1
	end_x = start_x + structure_dimensions[0] - 1
	return end_x, end_y, end_z


def get_facing_south_coordinates(start_x, start_y, start_z, structure_dimensions):
	end_z = start_z + structure_dimensions[0] - 1
	end_y = start_y + structure_dimensions[1] - 1
	end_x = start_x + structure_dimensions[2] - 1
	return end_x, end_y, end_z


def get_facing_west_coordinates(start_x, start_y, start_z, structure_dimensions):  # fixed
	end_z = start_z + structure_dimensions[2] - 1  # not subtracting makes it longer by one
	end_y = start_y + structure_dimensions[1] - 1
	end_x = start_x - structure_dimensions[0] + 1  # subtracting makes it longer by one.
	return end_x, end_y, end_z


def get_general_direction(player_yaw):
	specific_direction = get_specific_direction(player_yaw)
	north_dir = ["north", "north northeast", "north northwest", "northeast"]
	east_dir = ["east", "east northeast", "east southeast", "southeast"]
	south_dir = ["south", "south southeast", "south southwest", "southwest"]
	# west_dir = ["west", "west northwest", "west southwest", "northwest"]

	if specific_direction in north_dir:
		return "north"
	elif specific_direction in east_dir:
		return "east"
	elif specific_direction in south_dir:
		return "south"
	else:
		return "west"


def get_specific_direction(player_yaw):
	if player_yaw < 0:
		player_yaw += 360

	dir_int = int((player_yaw + 8) / 22.5)

	if dir_int == 0:
		return "west"
	elif dir_int == 1:
		return "west northwest"
	elif dir_int == 2: # west
		return "northwest"
	elif dir_int == 3:
		return "north northwest"
	elif dir_int == 4:
		return "north"
	elif dir_int == 5:
		return "north northeast"
	elif dir_int == 6: # north
		return "northeast"
	elif dir_int == 7:
		return "east northeast"
	elif dir_int == 8:
		return "east"
	elif dir_int == 9:
		return "east southeast"
	elif dir_int == 10: # east
		return "southeast"
	elif dir_int == 11:
		return "south southeast"
	elif dir_int == 12:
		return "south"
	elif dir_int == 13:
		return "south southwest"
	elif dir_int == 14:
		return "southwest"
	elif dir_int == 15:
		return "west southwest"
	else:
		return "west"


class PlayerPos(object):
	x = 0
	y = 0
	z = 0


if __name__ == '__main__':
	player_pos = PlayerPos()
	player_pos.x = 10
	player_pos.y = 10
	player_pos.z = 10

	get_build_coordinates(player_pos, 0, [3, 3, 3])

File: src/com/multicraft/CoordinateCalculations.java
package com.multicraft;

import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;

import org.bukkit.Location;

public class CoordinateCalculations {
	
	public static int[] getBuildCoordinates(Location loc, int[] dimensions) {
		int angle = (int) loc.getYaw();
		int startX = loc.getBlockX(), startY = loc.getBlockY(), startZ = loc.getBlockZ();
		String generalDirection = getGeneralDirection(angle);
		int[] dirs = new int[3];
		if(generalDirection.equals("north")) {
			dirs = getFacingNorthCoordinates(startX, startY, startZ, dimensions);
		}else if(generalDirection.equals("east")) {
			dirs = getFacingEastCoordinates(startX, startY, startZ, dimensions);
		}else if(generalDirection.equals("south")) {
			dirs = getFacingSouthCoordinates(startX, startY, startZ, dimensions);
		}else {
			dirs = getFacingWestCoordinates(startX, startY, startZ, dimensions);
		}
		return new int[] {dirs[0], dirs[1], dirs[2]};
	}
	
	private static int[] getFacingNorthCoordinates(int startX, int startY, int startZ, int[] dimensions) {
		// TODO Auto-generated method stub
		int endX = 0, endZ = 0, endY = 0;
		endZ = startZ - dimensions[0] + 1;
		endY = startY + dimensions[1] - 1;
		endX = startX - dimensions[2] + 1;
		return new int[] {endX, endY, endZ};
	}

	private static int[] getFacingEastCoordinates(int startX, int startY, int startZ, int[] dimensions) {
		// TODO Auto-generated method stub
		int endX = 0, endZ = 0, endY = 0;
		endZ = startZ - dimensions[2] + 1;
		endY = startY + dimensions[1] - 1;
		endX = startX + dimensions[0] - 1;
		return new int[] {endX, endY, endZ};
	}

	private static int[] getFacingSouthCoordinates(int startX, int startY, int startZ, int[] dimensions) {
		// TODO Auto-generated method stub
		int endX = 0, endZ = 0, endY = 0;
		endZ = startZ + dimensions[0] - 1;
		endY = startY + dimensions[1] - 1;
		endX = startX + dimensions[2] - 1;
		return new int[] {endX, endY, endZ};
	}

	private static int[] getFacingWestCoordinates(int startX, int startY, int startZ, int[] dimensions) {
		// TODO Auto-generated method stub
		int endX = 0, endZ = 0, endY = 0;
		endZ = startZ + dimensions[2] - 1;
		endY = startY + dimensions[1] - 1;
		endX = startX - dimensions[0] + 1;
		return new int[] {endX, endY, endZ};
	}

	private static String getGeneralDirection(int angle) {
		String specificDirection = getSpecificDirection(angle);
		Set<String> northDir = new HashSet<String>(Arrays.asList("north", "north northeast", "north northwest", "northeast"));
		Set<String> eastDir = new HashSet<String>(Arrays.asList("east", "east northeast", "east southeast", "southeast"));
		// Set<String> westDir = new HashSet<String>(Arrays.asList("north", "north northeast", "north northwest", "northeast"));
		Set<String> southDir = new HashSet<String>(Arrays.asList("south", "south southeast", "south southwest", "southwest"));
		
		if(northDir.contains(specificDirection)){
			return "north";
		}
		if(eastDir.contains(specificDirection)){
			return "east";
		}
		if(southDir.contains(specificDirection)){
			return "south";
		}
		return "west";
	}
	
	public static String getSpecificDirection(int angle) {
		if(angle < 0) {
			angle += 360;
		}
		int dirInt = (int) ((angle + 8) / 22.5);
		if(dirInt == 0)
			return "south";
		else if(dirInt == 1)
			return "south southwest";
		else if (dirInt == 2)
			return "southwest";
		else if(dirInt == 3)
			return "north southwest";
		else if(dirInt == 4)
			return "west"; // west
		else if(dirInt == 5)
			return "west northwest";
		else if(dirInt == 6)
			return "northwest";
		else if(dirInt == 7)
			return "north northwest";
		else if(dirInt == 8)
			return "north"; //north 
		else if(dirInt == 9)
			return "north northeast";
		else if(dirInt == 10)
			return "northeast";
		else if(dirInt == 11)
			return "south northeast";
		else if(dirInt == 12)
			return "east"; //east
		else if(dirInt == 13)
			return "east southeast";
		else if(dirInt == 14)
			return "southeast";
		else if(dirInt == 15)
			return "south southeast";
		else
			return "south"; // south
	}

}
